src/main/java/org/opentripplanner/routing/algorithm/raptor/transit/TripPatternForDate.java

```diff
@@ -57,6 +67,7 @@ public LocalDate getLocalDate() {
         if (o == null || getClass() != o.getClass())
             return false;
         TripPatternForDate that = (TripPatternForDate) o;
+        // FIXME equals method seems broken, it only checks the pattern ID
         return this.getTripPattern().getId() == that.getTripPattern().getId();
```

Comment:
I agree with everything you stated - this would have surprised me less if it used .equals() and .equals then compared the IDs, and included a comment explaining that IDs were unique identifiers.

Reply:
I agree with everything you stated - once you're giving objects IDs it's reasonable to use those as fast, unique identifiers. But I do think it's good to state on such methods whether a given class has identity equality, semantic equality, or equality based on enforced existence of only one instance with a particular ID, all of which have different meanings.